diff --git a/tboot-1.7.3/tboot/common/integrity.c b/tboot-1.7.3/tboot/common/integrity.c
index 066ba19..05f9a71 100644
--- a/tboot-1.7.3/tboot/common/integrity.c
+++ b/tboot-1.7.3/tboot/common/integrity.c
@@ -78,17 +78,20 @@ extern void apply_policy(tb_error_t error);
 
 #define EVTTYPE_TB_MEASUREMENT (0x400 + 0x101)
 extern bool evtlog_append(uint8_t pcr, tb_hash_t *hash, uint32_t type);
+extern bool read_policy_from_tpm(tpm_nv_index_t index,
+                void* policy_index, size_t *policy_index_size);
 
 typedef struct {
     uint8_t mac_key[VMAC_KEY_LEN/8];
     uint8_t shared_key[sizeof(_tboot_shared.s3_key)];
 } sealed_secrets_t;
 
+static void insert_geo_tag(void);
 
 static bool extend_pcrs(void)
 {
     tpm_pcr_value_t pcr17, pcr18;
-
+    
     tpm_pcr_read(2, 17, &pcr17);
     tpm_pcr_read(2, 18, &pcr18);
     printk(TBOOT_DETA"PCRs before extending:\n");
@@ -111,10 +114,42 @@ static bool extend_pcrs(void)
     printk(TBOOT_DETA"PCRs after extending:\n");
     printk(TBOOT_DETA"  PCR 17: "); print_hash((tb_hash_t *)&pcr17, TB_HALG_SHA1);
     printk(TBOOT_DETA"  PCR 18: "); print_hash((tb_hash_t *)&pcr18, TB_HALG_SHA1);
-
+    insert_geo_tag(); /*insert geo tag */
+   
     return true;
 }
 
+static void insert_geo_tag(void)
+{
+    tpm_pcr_value_t pcr22;
+    size_t geo_tag_size = 20; /* standard hash for now */
+    unsigned char geo_tag[20] = {'0'};
+
+    printk(TBOOT_DETA"reading GEO TAG from from TPM NV_Index %u...\n",
+            TPM_NV_INDEX_GEO);
+
+    if ( !read_policy_from_tpm(TPM_NV_INDEX_GEO,
+                geo_tag, &geo_tag_size) ) {
+        printk(TBOOT_DETA"\t:reading GEO TAG failed from NV\n");
+        return;
+    } 
+
+    tpm_pcr_read(2, 22, &pcr22);
+
+    printk(TBOOT_DETA"PCR22 before GEO-Tag extending:\n");
+    printk(TBOOT_DETA"  PCR 22: "); print_hash((tb_hash_t *)&pcr22, TB_HALG_SHA1);
+    printk(TBOOT_DETA"\t:%lu GEO NV bytes read\n", geo_tag_size);
+
+    memcpy((void*)(pcr22.digest), geo_tag, 20); 
+    if ( tpm_pcr_extend(2, 22, &pcr22, NULL) != TPM_SUCCESS ) {
+        printk(TBOOT_DETA"\t:reading GEO TAG PRC22 Extend failed \n");        
+    }
+
+    tpm_pcr_read(2, 22, &pcr22);
+    printk(TBOOT_DETA"PCR22 after GEO-TAG extending:\n");
+    printk(TBOOT_DETA"  PCR 22: "); print_hash((tb_hash_t *)&pcr22, TB_HALG_SHA1);
+    return; 
+}
 static void print_pre_k_s3_state(void)
 {
     printk(TBOOT_DETA"pre_k_s3_state:\n");
diff --git a/tboot-1.7.3/tboot/common/policy.c b/tboot-1.7.3/tboot/common/policy.c
index e8a7704..c3463d3 100644
--- a/tboot-1.7.3/tboot/common/policy.c
+++ b/tboot-1.7.3/tboot/common/policy.c
@@ -184,7 +184,7 @@ static const tb_policy_t* g_policy = &_def_policy;
  *
  * policy_index_size is in/out
  */
-static bool read_policy_from_tpm(tpm_nv_index_t index,
+bool read_policy_from_tpm(tpm_nv_index_t index,
                 void* policy_index, size_t *policy_index_size)
 {
 #define NV_READ_SEG_SIZE    256
diff --git a/tboot-1.7.3/tboot/include/tpm.h b/tboot-1.7.3/tboot/include/tpm.h
index 67b42d4..4e22df9 100644
--- a/tboot-1.7.3/tboot/include/tpm.h
+++ b/tboot-1.7.3/tboot/include/tpm.h
@@ -51,6 +51,7 @@
 
 #define TPM_NR_LOCALITIES             5
 
+#define TPM_NV_INDEX_GEO              ((uint32_t)0x00000005)
 /*
  * return code:
  * The TPM has five types of return code. One indicates successful operation
